pennylane/resources/second_quantization.py

```diff
@@ -15,374 +15,529 @@
+        >>> alpha = 10
+        >>> beta = 20
+        >>> unitary_cost(n, rank_r, rank_m, rank_max, br, alpha, beta)
+        2007
+        """
+        if n <= 0 or not isinstance(n, int) or n % 2 != 0:
+            raise ValueError("The number of spin-orbitals must be a positive even integer.")
 
-    if rank_r <= 0 or not isinstance(rank_r, int):
-        raise ValueError("The rank of the first factorization step must be a positive integer.")
+        if rank_r <= 0 or not isinstance(rank_r, int):
+            raise ValueError("The rank of the first factorization step must be a positive integer.")
 
-    if rank_m <= 0:
-        raise ValueError("The rank of the second factorization step must be a positive number.")
+        if rank_m <= 0:
+            raise ValueError("The rank of the second factorization step must be a positive number.")
 
-    if rank_max <= 0 or not isinstance(rank_max, int):
-        raise ValueError(
-            "The maximum rank of the second factorization step must be a positive" " integer."
-        )
+        if rank_max <= 0 or not isinstance(rank_max, int):
+            raise ValueError(
+                "The maximum rank of the second factorization step must be a positive" " integer."
+            )
 
-    if br <= 0 or not isinstance(br, int):
-        raise ValueError("br must be a positive integer.")
+        if br <= 0 or not isinstance(br, int):
+            raise ValueError("br must be a positive integer.")
+
+        if alpha <= 0 or not isinstance(alpha, int):
+            raise ValueError("alpha must be a positive integer.")
 
-    if alpha <= 0 or not isinstance(alpha, int):
-        raise ValueError("alpha must be a positive integer.")
+        if beta <= 0 or not isinstance(beta, int):
+            raise ValueError("beta must be a positive integer.")
 
-    if beta <= 0 or not isinstance(beta, int):
-        raise ValueError("beta must be a positive integer.")
+        rank_rm = rank_r * rank_m
+
+        # eta is computed based on step 1.(a) in page 030305-41 of PRX Quantum 2, 030305 (2021)
+        eta = np.array([np.log2(n) for n in range(1, rank_r + 1) if rank_r % n == 0])
+        eta = int(np.max([n for n in eta if n % 1 == 0]))
+
+        nxi = np.ceil(np.log2(rank_max))  # Eq. (C14) of PRX Quantum 2, 030305 (2021)
+        nl = np.ceil(np.log2(rank_r + 1))  # Eq. (C14) of PRX Quantum 2, 030305 (2021)
+        nlxi = np.ceil(np.log2(rank_rm + n / 2))  # Eq. (C15) of PRX Quantum 2, 030305 (2021)
+
+        bp1 = nl + alpha  # Eq. (C27) of PRX Quantum 2, 030305 (2021)
+        bo = nxi + nlxi + br + 1  # Eq. (C29) of PRX Quantum 2, 030305 (2021)
+        bp2 = nxi + alpha + 2  # Eq. (C31) of PRX Quantum 2, 030305 (2021)
+
+        # cost is computed using Eq. (C39) of PRX Quantum 2, 030305 (2021)
+        cost = (
+            9 * nl - 6 * eta + 12 * br + 34 * nxi + 8 * nlxi + 9 * alpha + 3 * n * beta - 6 * n - 43
+        )
+        cost += self._qrom_cost((rank_r, 1, 0, bp1, -1))[0]
```

Comment:
Maybe we can add inline comments here to show what type of cost is added in each step. That would make code more readable.